rclcpp_lifecycle/src/transition.cpp

```diff
@@ -36,54 +50,109 @@ Transition::Transition(uint8_t id, const std::string & label)
 
 Transition::Transition(
   uint8_t id, const std::string & label,
-  State && start, State && goal)
-: owns_rcl_transition_handle_(true)
+  State && start, State && goal,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_transition_handle_(true),
+  transition_handle_(nullptr)
 {
-  auto transition_handle = new rcl_lifecycle_transition_t;
+  auto transition_handle = reinterpret_cast<rcl_lifecycle_transition_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_transition_t), allocator_.state));
+  if (!transition_handle) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_transition_t");
+  }
   transition_handle->id = id;
-  transition_handle->label = label.c_str();
+  transition_handle->label = rcutils_strndup(
+    label.c_str(), label.size(), allocator_);
 
-  auto start_state = new rcl_lifecycle_state_t;
+  auto start_state = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
+  if (!start_state) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
+  }
   start_state->id = start.id();
-  start_state->label = start.label().c_str();
+  start_state->label =
+    rcutils_strndup(start.label().c_str(), start.label().size(), allocator_);
 
-  auto goal_state = new rcl_lifecycle_state_t;
+  auto goal_state = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
+  if (!goal_state) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
+  }
   goal_state->id = goal.id();
-  goal_state->label = start.label().c_str();
+  goal_state->label =
+    rcutils_strndup(goal.label().c_str(), goal.label().size(), allocator_);
 
   transition_handle->start = start_state;
   transition_handle->goal = goal_state;
   transition_handle_ = transition_handle;
 }
 
-Transition::Transition(const rcl_lifecycle_transition_t * rcl_lifecycle_transition_handle)
-: owns_rcl_transition_handle_(false)
+Transition::Transition(
+  const rcl_lifecycle_transition_t * rcl_lifecycle_transition_handle,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_transition_handle_(false),
+  transition_handle_(nullptr)
 {
-  transition_handle_ = rcl_lifecycle_transition_handle;
+  transition_handle_ = const_cast<rcl_lifecycle_transition_t *>(rcl_lifecycle_transition_handle);
 }
 
 Transition::~Transition()
 {
-  if (owns_rcl_transition_handle_) {
-    if (transition_handle_->start) {
-      delete transition_handle_->start;
+  if (!transition_handle_) {
```

Comment:
Why does the logic in this destructor different from the one for the `State` class?. E.g. it shouldn't be possible for `transition_handle_` to be null.